File: src/constants/style-filters.ts
export const SEASON_OPTIONS = ['spring', 'summer', 'fall', 'winter', 'all-season'] as const;

/**
 * Occasions are derived from a garment's type rather than tagged by hand (see
 * utils/garment-occasions). The list is therefore limited to what a garment's
 * type can actually imply -- "party" and "travel" used to be options here, but
 * nothing about a garment tells you it is for travel, so filtering by them
 * could only ever return nothing. These five match the Occasion domain type.
 */
export const OCCASION_OPTIONS = ['casual', 'work', 'formal', 'sport', 'lounge'] as const;

export type SeasonOption = (typeof SEASON_OPTIONS)[number];
export type OccasionOption = (typeof OCCASION_OPTIONS)[number];

type StyleFilterOption = SeasonOption | OccasionOption;

export const STYLE_FILTER_EMOJIS: Record<StyleFilterOption, string> = {
  spring: '🌸',
  summer: '☀️',
  fall: '🍂',
  winter: '❄️',
  'all-season': '🧥',
  casual: '😌',
  work: '💼',
  formal: '🎩',
  sport: '🏃',
  lounge: '🛋️',
};

export const STYLE_FILTER_COLORS: Record<StyleFilterOption, string> = {
  spring: '#F472B6',
  summer: '#F59E0B',
  fall: '#FB7185',
  winter: '#60A5FA',
  'all-season': '#14B8A6',
  casual: '#10B981',
  work: '#6366F1',
  formal: '#334155',
  sport: '#F43F5E',
  lounge: '#A855F7',
};

File: domain/src/main/kotlin/com/wardrobapp/domain/StyleFilters.kt
package com.wardrobapp.domain

/**
 * Seasons a garment can be tagged with, and that a suggestion can be filtered
 * by. `tag` is the lowercase string stored in a garment's tags.
 */
enum class Season(val tag: String) {
    SPRING("spring"),
    SUMMER("summer"),
    FALL("fall"),
    WINTER("winter"),
    ALL_SEASON("all-season");

    /** The season that argues against this one, or null for ALL_SEASON. */
    val opposite: Season?
        get() = when (this) {
            SPRING -> FALL
            SUMMER -> WINTER
            FALL -> SPRING
            WINTER -> SUMMER
            ALL_SEASON -> null
        }

    companion object {
        fun fromTag(tag: String): Season? = entries.find { it.tag == tag.lowercase().trim() }
    }
}

/**
 * Occasions a garment's *type* can imply.
 *
 * Occasion is derived from the garment's type rather than tagged by hand, so the
 * list is limited to what a type can actually tell you -- "party" and "travel"
 * were dropped upstream because nothing about a garment implies them, which made
 * filtering by them return nothing at all.
 *
 * Declaration order is the order results are reported in.
 */
enum class Occasion(val id: String) {
    CASUAL("casual"),
    WORK("work"),
    FORMAL("formal"),
    SPORT("sport"),
    LOUNGE("lounge");

    companion object {
        fun fromId(id: String): Occasion? = entries.find { it.id == id.lowercase().trim() }
    }
}

/**
 * The season a date falls in.
 *
 * The month arrives as an argument rather than being read from a clock, which is
 * the same reason the suggestion engine takes its randomness that way: a
 * suggestion run has to be reproducible. Zero-based to match what both
 * `Date.getMonth()` and `Calendar.MONTH` hand over, so neither caller has to
 * remember to shift it.
 *
 * Never returns [Season.ALL_SEASON]: that is a property a garment can have, not
 * a time of year it can be.
 */
fun seasonOfMonth(monthZeroBased: Int): Season = when (monthZeroBased) {
    in 2..4 -> Season.SPRING
    in 5..7 -> Season.SUMMER
    in 8..10 -> Season.FALL
    else -> Season.WINTER
}

/** What the user asked for, if anything. */
data class SuggestionPreferences(
    val seasons: List<Season> = emptyList(),
    val occasion: Occasion? = null,
)
